api/operator/v1alpha1/istiocsr_types.go

```diff
@@ -0,0 +1,246 @@
+package v1alpha1
+
+import (
+	certmanagerv1 "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+func init() {
+	SchemeBuilder.Register(&IstioCSR{}, &IstioCSRList{})
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+//+kubebuilder:object:root=true
+
+// IstioCSRList is a list of IstioCSR objects.
+type IstioCSRList struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard list's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ListMeta `json:"metadata"`
+	Items           []IstioCSR `json:"items"`
+}
+
+// +genclient
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// +kubebuilder:subresource:status
+
+// IstioCSR describes configuration and information about the managed istio-csr
+// agent. The name must be `default`.
+//
+// When an IstioCSR is created, a new deployment is created which manages the
+// istio-csr agent and keeps it in the desired state.
+// +kubebuilder:object:root=true
+// +kubebuilder:printcolumn:name="Age",type="date",JSONPath=".metadata.creationTimestamp"
+// +kubebuilder:printcolumn:name="GRPC Endpoint",type="string",JSONPath=".status.istioCSRGRPCEndpoint"
+// +kubebuilder:validation:XValidation:rule="self.metadata.name == 'default'",message="istiocsr is a singleton, .metadata.name must be 'default'"
+// +operator-sdk:csv:customresourcedefinitions:displayName="IstioCSR"
+type IstioCSR struct {
+	metav1.TypeMeta `json:",inline"`
+
+	// metadata is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#metadata
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// spec is the specification of the desired behavior of the IstioCSR.
+	Spec IstioCSRSpec `json:"spec,omitempty"`
+
+	// status is the most recently observed status of the IstioCSR.
+	Status IstioCSRStatus `json:"status,omitempty"`
+}
+
+// IstioCSRSpec is the specification of the desired behavior of the IstioCSR.
+type IstioCSRSpec struct {
+	// istioCSRConfig is for configuring the istio-csr agent behavior.
+	IstioCSRConfig *IstioCSRConfig `json:"istioCSRConfig,omitempty"`
+
+	// controllerConfig is for configuring the controller for setting up
+	// defaults to enable istio-csr agent.
+	ControllerConfig *ControllerConfig `json:"controllerConfig,omitempty"`
+}
+
+// IstioCSRConfig is for configuring the istio-csr agent behavior.
+type IstioCSRConfig struct {
+	// logLevel is for setting verbosity of istio-csr agent logging.
+	// Supported log levels: 1-5.
+	// +kubebuilder:default:=1
+	// +kubebuilder:validation:Minimum:=1
+	// +kubebuilder:validation:Maximum:=5
+	// +optional
+	LogLevel int32 `json:"logLevel,omitempty"`
```

Comment:
Could we please be consistent with the other logLevels, for UX parity.
Refer: https://github.com/openshift/cert-manager-operator/blob/master/pkg/controller/deployment/deployment_log_level.go i.e. use `{Normal, Debug, Trace, TraceAll}`.

Reply:
Log level here is for operand, and I think keeping it same as defined by operand would be better, otherwise we need to document the mapping for first. Else we could also remove this option from user config and controller can set the same defined for the operator.

Reply:
yes, we should document it as we already do in https://docs.openshift.com/container-platform/4.17/security/cert_manager_operator/cert-manager-log-levels.html for the cert-manager operand resources.

For the other part, I'd prefer keeping the logLevel in the istio_csr resource itself because by design the way everything is implemented/proposed right now it seems to me we should prefer separation of `certmanager.operator.openshift.io` and `istiocsr.openshift.io` in their respective jobs of reconciling their own operand deployments exclusively.